[Core][Frontend] Support Passing Multimodal Processor Kwargs

alex-jw-brooks opened the pull request:
In support of https://github.com/vllm-project/vllm/issues/7861 - It's split into two PRs to make it easier to review

This PR adds support for passing `processor_kwargs` at initialization time to override values, e.g., in the image processor config and adds a pattern that we can use to easily implement such overrides, especially in the case of multimodal models. More specifically, it adds a well-defined way to pass valid `processor_kwargs` as expanded keyword-only arguments (which should have default values) to:

- The input mapper (e.g,. default mapper wrapping the image processor from HF, or a custom mapper implemented in the model class)
- The input processor
- Max token calculations
- Dummy data for memory profiling

This is important since the provided overrides may greatly change the number of image tokens per multimodal instance.

It also forwards some args to the video processor, since those weren't being pushed through yet.
### Some Implementation Considerations
I've tried to implement this to be easily extensible for potential inference-time configuration and intuitive to use coming from `transformers`. More specifically, I was aiming for:

- avoid writing any configs on disk (since that's the main reason we need this anyway)
- avoid mutating any configs in-memory
- Make no assumptions about config structure are made; having a user specify where the value of something like `num_crops` comes from is a bad experience, and implementing a per-model mapping of allowed config overrides is also fragile since things could be moved around a bit for models that are written to try to generically support fine-tuned models with different LLMs, etc
- Assume that most overrides are going to be added gradually added; especially for the mapper, since it is a VLLM model implementation detail of whether or not it's using a custom mapper or falling back to something like the HF image processor. If a processor_kwarg is not supported, it will `warn` that it'll be unused and drop it from the processor kwargs, instead of throwing some nasty unexpected keyword argument deep in the model class
- Keep compatibility with the existing interfaces for max_token_counts, mapper, input processor, etc
- Should be easy for us to check if something has already been added for a model

### Examples
I've opened a second PR on top of this PR which makes `num_crops` at init time for phi3v models which illustrates the usage [here](https://github.com/vllm-project/vllm/pull/8658); see [this commit](https://github.com/vllm-project/vllm/pull/8658/commits/adde77635fb2d3a51324a39a5a6a216245d24686) for what adding a `processor_kwarg` looks like, as well as example tests for each place it's used in phi3v.


You can also try running those examples on this to see what happens if it's unsupported. Since it leverages the default mapper, whose kwargs can't easily be inspected since it gets created through the automodel, there will be some mismatches that eventually crash since the processor kwargs *will* be used to initialize the auto model, but you'll at least get a bunch of things like:
```
The following intended overrides are not keyword-only args and and will be dropped: {'num_crops'}
```
out in the warmup. In situations where the class has its own mapper, and a bad kwarg was used, this would be ignored like it is everywhere else (same case as the sad not found kwargs tests here)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

alex-jw-brooks committed:
Allow for processor kwarg overrides

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Pass processor through to partial

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add default & processor kwarg override tests

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Don't allow ctx or inputs as kwargs

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add kwarg override for processor to dummy data factories

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add kwarg override forr processor to max token calc

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Move kwarg only override func to utils

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Force processor kwargs to be keyword-only

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Pass unfiltered processor kwargs to default mapper

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add hack for mapper preprocessor kwargs

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Move context builder to test utils

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Simplify dummy data processor kwarg & add tests

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add tests for max multimodal token kwarg overrides

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Format registry

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Fix default mapper comparison

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Move kwarg filtering into hf processor getter

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Enable processor_kwargs in video processor

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add tests for mapper processor_kwargs

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Update mapper not on multimodal processor kwargs

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
processor kwarg test cleanup

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Use common context builder in processor kwarg tests

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

DarkLight1337 commented on vllm/inputs/registry.py:
```diff
@@ -235,29 +241,55 @@ def wrapper(model_cls: N) -> N:
-    def process_input(self, model_config: "ModelConfig",
-                      inputs: LLMInputs) -> LLMInputs:
+    def _process_input(self, inputs: LLMInputs, model_config: "ModelConfig",
+                       processor: Callable, **processor_kwargs) -> LLMInputs:
```
```suggestion
                       processor: InputProcessor, **processor_kwargs: Any) -> LLMInputs:
```

DarkLight1337 commented on vllm/inputs/registry.py:
```diff
@@ -235,29 +241,55 @@ def wrapper(model_cls: N) -> N:
+        return processor(InputContext(model_config), inputs,
+                         **processor_kwargs)
+
+    def create_input_processor(self, model_config: "ModelConfig") -> Callable:
```
```suggestion
    def create_input_processor(self, model_config: "ModelConfig"):
```

DarkLight1337 replied:
If you do want to explicitly specify it, it should return `Callable[[LLMInputs], LLMInputs]` here.

DarkLight1337 commented on vllm/inputs/registry.py:
```diff
@@ -235,29 +241,55 @@ def wrapper(model_cls: N) -> N:
+                                 **processor_kwargs)
+
+    def _get_model_input_processor(self,
+                                   model_config: "ModelConfig") -> Callable:
```
```suggestion
                                   model_config: "ModelConfig"):
```

DarkLight1337 replied:
If you do want to explicitly specify it, it should return `InputProcessor` here.

DarkLight1337 commented on vllm/inputs/registry.py:
```diff
@@ -211,7 +217,7 @@ def _default_input_processor(self, ctx: InputContext,
         return inputs
 
-    def register_input_processor(self, processor: InputProcessor):
+    def register_input_processor(self, processor: InputProcessor) -> Callable:
```
If you do want to explicitly specify it, it should return `Callable[[N], N]` here.

DarkLight1337 reviewed (commented):
Hmm, originally I was thinking of dynamic kwargs via multi-modal inputs, e.g.

```py
llm.generate({"multi_modal_data": {"image": {"data": image, "options": image_kwargs}}})
```

but this works too!

I'll take a closer look later today.

alex-jw-brooks commented:
Sounds good, thanks @DarkLight1337! 🙂 I wanted to open this one first to since it's useful by itself, and to make sure the overall approach looks reasonable and had good test coverage since it's touching a few different things.

I'm happy to add that in a follow-up PR though - I did implement this with that in mind, and it won't require changes to individual models to support it if additional init time processor kwargs are exposed in the meantime! 🤞

DarkLight1337 commented on vllm/inputs/registry.py:
```diff
@@ -73,12 +74,16 @@ def __call__(
 
         Note:
             :data:`InputProcessor` is not applied to the dummy data.
+            The processor_kwargs are overrides provided at initialization
+            time to values in the config whose values may affect the number
+            of tokens per instance.
```
```suggestion

            The :code:`processor_kwargs` are overrides provided at initialization
            time to values in the config whose values may affect the number
            of tokens per instance.
```

Add a newline to separate the two sentences by paragraph.

DarkLight1337 reviewed (commented):
Overall looks good, thanks for putting so much effort into the tests! Just some minor fixes to do.

alex-jw-brooks committed:
Update vllm/entrypoints/llm.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

alex-jw-brooks committed:
Update vllm/inputs/registry.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

alex-jw-brooks committed:
Update vllm/inputs/registry.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

alex-jw-brooks committed:
Update vllm/inputs/registry.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

alex-jw-brooks committed:
Update vllm/inputs/registry.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

alex-jw-brooks committed:
Update vllm/inputs/registry.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

alex-jw-brooks committed:
Fix formatting

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks commented:
Awesome, thanks for the quick review @DarkLight1337! I went ahead and merged your suggestions and reran formatting 🙂

DarkLight1337 commented on vllm/config.py:
```diff
@@ -122,6 +122,8 @@ class ModelConfig:
             can not be gathered from the vllm arguments. 
         config_format: The config format which shall be loaded.
             Defaults to 'auto' which defaults to 'hf'.
+        processor_kwargs: Arguments to be forwarded to the model's processor,
```
From my understanding, this only applies to multimodal processors at the moment, not the tokenizer. If that's the case, let's rename this to `mm_processor_kwargs` to make that more explicit

alex-jw-brooks replied:
Sounds good! I've renamed `processor_kwargs` -> `mm_processor_kwargs` and updated the engine CLI arg to make it more clear that it's for multimodal!

alex-jw-brooks committed:
Rename processor kwargs to mm processor kwargs

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

DarkLight1337 commented on vllm/config.py:
```diff
@@ -122,6 +122,8 @@ class ModelConfig:
             can not be gathered from the vllm arguments. 
         config_format: The config format which shall be loaded.
             Defaults to 'auto' which defaults to 'hf'.
+        mm_processor_kwargs: Arguments to be forwarded to the model's processor,
+            e.g., tokenizer, image processor, or custom processor callable.
```
```suggestion
        mm_processor_kwargs: Arguments to be forwarded to the model's processor
            for multi-modal data, e.g., image processor.
```

DarkLight1337 committed:
Update docstring

DarkLight1337 reviewed (approved):
LGTM now, thanks for implementing this!

DarkLight1337 committed:
Merge branch 'main' into support_processor_kwargs

DarkLight1337 committed:
Try to fix CUDA reinitialization error

DarkLight1337 commented:
Looks like there is CUDA reinitialization error after switching over the imports... @youkaichao can you help figure out how this is happening?

youkaichao commented:
this is a large pr. which change do you mean?

DarkLight1337 commented:
I suspect it has something to do with the use of `inspect` in `utils.py`, honestly not sure.

DarkLight1337 commented:
Could you remind me of the procedure used to check which module initialized CUDA?

DarkLight1337 committed:
Consolidate processor loading

DarkLight1337 commented:
Nvm, I figured it out. Just need to add a breakpoint in `torch.cuda._lazy_init`. It looks like the culprit is `InputRegistry._get_model_input_processor` which imports the model file.

DarkLight1337 committed:
Fix CUDA reinitialization error

DarkLight1337 committed:
Simplify code

DarkLight1337 committed:
Fix tests

alex-jw-brooks commented:
Thanks a lot for the support @DarkLight1337! I remember you had also fixed another issue in one of my other PRs around CUDA reinitialization for Qwen-VL tests in the multi-gpu weight loading. Do you have any suggestions for avoiding these sorts of issues moving forward, e.g., at what point in the initialization process in VLLM that it's okay to import things that need to have CUDA initialized? I assume this is done explicitly somewhere, since AFAIK things like `torch.cuda.init()` are idempotent.

I also added a follow-up issue to this one around request-time `mm_processor_kwargs` support, which hopefully should be pretty easy: https://github.com/vllm-project/vllm/issues/8742

DarkLight1337 commented:
Basically we should avoid importing model files before starting the vLLM engine.

alex-jw-brooks commented:
Awesome, thanks! I'll keep that in mind in future PRs
